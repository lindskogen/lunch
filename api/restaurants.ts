import { NowRequest, NowResponse } from "@now/node";
import { fetchAll } from "../lib/lunch-fetcher/fetch-restaurants";

export default async (req: NowRequest, res: NowResponse) => {
  const restaurants = await fetchAll();

  res.setHeader("Cache-Control", "s-maxage=3600, stale-while-revalidate");

  res.json({ restaurants, lastUpdated: new Date().toISOString() });
};
