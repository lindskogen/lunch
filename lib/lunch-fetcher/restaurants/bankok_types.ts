export interface Ingestion {
  price: number;
  type: string;
}

export interface Item {
  id: string;
  name: string;
  preview: string;
  info: string;
  description: string;
  outOfStock: boolean;
  sortOrder: number;
  category: number;
  additionalItems: any[];
  ingestions: Ingestion[];
  itemCategory: string;
}

export interface ItemCategory {
  id: string;
  name: string;
  preview?: any;
  description: string;
  category: number;
}

export interface RestaurantData {
  items: Item[];
  itemCategories: ItemCategory[];
}
