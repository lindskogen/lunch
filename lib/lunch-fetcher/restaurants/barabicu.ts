import * as _ from "lodash";
import { isNotNull } from "../lib/utils";
import { FoodItem, Restaurant, WeekDay, RestaurantDayMenu } from "../types";

export const url = "https://barabicu.se/";

const weekdayTranslations: WeekDay[] = [
  "Måndag",
  "Tisdag",
  "Onsdag",
  "Torsdag",
  "Fredag"
];
export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $("#todays-lunch li")
    .toArray()
    .map<RestaurantDayMenu>((dayNode, index) => {
      return {
        items: $("h3", dayNode)
          .toArray()
          .map<FoodItem | null>(elem => {
            if ($(elem).text() === '' || $(elem.next).text() === '') {
              return null;
            }
            return {
              name: $(elem.next).text(),
              title: $(elem).text().split('•')[0].trim()
            };
          }).filter(isNotNull),
        wday: weekdayTranslations[index]
      };
    });

  return { name: "Barabicu", url, days: menus };
};
