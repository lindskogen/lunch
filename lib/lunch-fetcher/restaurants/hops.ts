import * as _ from "lodash";
import { id, isNotNull, isNotZero, isWeekday } from "../lib/utils";
import { FoodItem, Restaurant, WeekDay } from "../types";

export const url = "http://www.hopsbar.se/";

export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $("#Veckansmeny")
    .toArray()
    .map(item => {
      const allNodes = _.flatMap(item.childNodes, node =>
        $(node)
          .text()
          .trim()
          .split(/\n/)
      )
        .map(s => s.trim())
        .filter(id);

      const weekMealsStartIndex = allNodes.findIndex(
        text => text === "Veckans rätter"
      );

      const weekItems: FoodItem[] = _.takeWhile(
        allNodes.slice(weekMealsStartIndex + 1),
        text => !_.includes(text, " kaffe ")
      ).map(food => {
        const foodname = food.trim();
        const matches = foodname.split(':');

        if (matches.length > 1) {
          const [title, name] = matches;
          return { name: name.trim(), title: title.trim() };
        }

        return { name: foodname };
      });

      const splitIndexes = allNodes
        .map((text, index) => (isWeekday(text) ? index : null))
        .filter(isNotNull);

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map(splitIndex => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return days
        .filter(([wday]) => isWeekday(wday))
        .map(([wday, ...foods]) => {
          const items = foods.map(food => ({ name: food.trim() }));

          return {
            wday: wday as WeekDay,
            items: _.takeWhile(
              items,
              ({ name }) => name !== "Veckans rätter"
            ).concat(weekItems)
          };
        });
    });

  return { name: "Hops", url, days: _.flatten(menus) };
};
