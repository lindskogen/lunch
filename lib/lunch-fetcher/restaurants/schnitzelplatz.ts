import {take, drop} from "lodash-es";
import { Restaurant, FoodItem, RestaurantDayMenu, WeekDay } from "../types";
import { CheerioAPI } from "cheerio";

export const url = "https://schnitzelplatz.se/lunch/";

const firstPartOfSplitOrAll = (str: string, split: string): string => {
  const parts = str.split(split);
  if (parts && parts.length > 1) {
    return parts[0];
  } else {
    return str;
  }
};

export const parseHtml = ($: CheerioAPI): Restaurant => {
  const menus = $(".introbox__container h4")
    .get()
    .map(node => [
      $(node).text().trim(),
      firstPartOfSplitOrAll(
        $(node)
          .next("p")
          .text(),
        "\n"
      )
    ]);

  const numWeekItems = menus.length - 5;

  const weekItems: FoodItem[] = take(menus, numWeekItems).map(
    ([title, name]) => ({
      title: title.toUpperCase(),
      name
    })
  );
  const dayItems = drop(menus, numWeekItems);

  const days: RestaurantDayMenu[] = dayItems.map(([day, name]) => ({
    wday: day as WeekDay,
    items: [{ name }].concat(weekItems)
  }));

  return { name: "Schnitzelplatz Lagerhuset", url, days };
};
