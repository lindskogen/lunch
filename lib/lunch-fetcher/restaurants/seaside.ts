import * as _ from "lodash";
import { isWeekday, isNotNull, isNotZero } from "../lib/utils";

export const url = "http://seaside.kvartersmenyn.se/";

export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $(".day .meny")
    .toArray()
    .map(item => {
      const allNodes = item.childNodes
        .filter(i => i.tagName === "strong" || i.nodeValue)
        .map(item => $(item).text());

      const splitIndexes = allNodes
        .map((text, index) => (isWeekday(text) ? index : null))
        .filter(isNotNull);

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map(splitIndex => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return days
        .filter(([wday]) => isWeekday(wday))
        .map(([wday, ...foods]) => {
          const foodsRegex = /^([A-Ã–]+( [0-9])?) ?[\n :]([^0-9:\n]+)/gm;

          const allFoodsForWeekDay = foods.join("\n");

          const parsedFoods = [];

          let res = foodsRegex.exec(allFoodsForWeekDay);
          while (res !== null) {
            const [, title, , name] = res;
            parsedFoods.push({ title, name: name.trim() });
            res = foodsRegex.exec(allFoodsForWeekDay);
          }

          return {
            wday: wday as WeekDay,
            items: parsedFoods
          };
        });
    });

  return { name: "Seaside", url, days: _.flatten(menus) };
};
