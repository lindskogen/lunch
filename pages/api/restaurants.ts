import { NowRequest, NowResponse } from "@now/node";
import { fetchAll } from "../../lib/lunch-fetcher/fetch-restaurants";

const restaurants = async (req: NowRequest, res: NowResponse) => {
  const restaurants = await fetchAll();

  res.setHeader("Cache-Control", "s-maxage=3600, stale-while-revalidate");

  res.json({ restaurants, lastUpdated: new Date().toISOString() });
};

export default restaurants;
