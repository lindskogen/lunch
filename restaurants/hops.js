const _ = require("lodash");
const { isWeekday, isNotNull, isNotZero, id } = require("../lib/utils");

module.exports.parseHtml = $ => {
  const menus = $("#Veckansmeny")
    .toArray()
    .map(item => {
      const allNodes = _.flatMap(item.childNodes, node =>
        $(node)
          .text()
          .trim()
          .split(/\n/)
      ).filter(id);

      const splitIndexes = allNodes
        .map((text, index) => (isWeekday(text) ? index : null))
        .filter(isNotNull);

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map(splitIndex => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return days
        .filter(([wday]) => isWeekday(wday))
        .map(([wday, ...foods]) => ({
          wday,
          items: _.take(foods.map(food => ({ name: food.trim() })), 2)
        }));
    });

  return { name: "Hops", days: menus };
};
