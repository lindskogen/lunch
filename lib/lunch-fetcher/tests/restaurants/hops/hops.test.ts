import * as cheerio from "cheerio";
import { parseHtml } from "../../../restaurants/hops";

import { mock as mock1 } from "./mock1";

describe("hops", () => {
  describe("when provided with mock 1", () => {
    const result = parseHtml(cheerio.load(mock1));
    it("should work", () => {
      expect(result).toMatchSnapshot();
    });

    it("should parse 5 days", () => {
      expect(result.days).toHaveLength(5);
    });

    it("should remove colon from title and name", () => {
      result.days.forEach((day) => {
        day.items.forEach((item) => {
          expect(item.name).not.toMatch(/:/);
        });
      });
    });

    it("should have no empty names", () => {
      result.days.forEach((day) => {
        day.items.forEach((item) => {
          expect(item.name).not.toHaveLength(0);
        });
      });
    });
  });
});
