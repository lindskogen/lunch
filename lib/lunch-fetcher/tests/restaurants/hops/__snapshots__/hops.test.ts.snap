// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hops when provided with mock 1 should work 1`] = `
Object {
  "days": Array [
    Object {
      "items": Array [
        Object {
          "name": "Klassisk Wallenbergare med skirat smör samt potatismos.",
        },
        Object {
          "name": "Panerad vittring med kräft- och västerbottenkräm, citron samt kokt potatis och citron",
        },
        Object {
          "name": "Med köttfärs, bacon, champinjoner och svartpeppar.",
          "title": "Pasta",
        },
        Object {
          "name": "Pannbiff med brynt lök, potatismos och lingonsylt samt gräddsås.",
          "title": "Husman",
        },
        Object {
          "name": "Chilibiff med rostade rotfrukter samt paprikasås.",
          "title": "Vegetariskt",
        },
        Object {
          "name": "Välj mellan kyckling eller hand skalade räkor med krutonger, bacon, parmesanost & ceasardressing.",
          "title": "Caesarsallad",
        },
      ],
      "wday": "Måndag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Stekt fläsk med raggmunkar och lingon alternativt löksås och kokt potatis.",
        },
        Object {
          "name": "Kokt torskfilé med citron och salladslöksås serveras med chili frästa räkor samt potatismos",
        },
        Object {
          "name": "Med köttfärs, bacon, champinjoner och svartpeppar.",
          "title": "Pasta",
        },
        Object {
          "name": "Pannbiff med brynt lök, potatismos och lingonsylt samt gräddsås.",
          "title": "Husman",
        },
        Object {
          "name": "Chilibiff med rostade rotfrukter samt paprikasås.",
          "title": "Vegetariskt",
        },
        Object {
          "name": "Välj mellan kyckling eller hand skalade räkor med krutonger, bacon, parmesanost & ceasardressing.",
          "title": "Caesarsallad",
        },
      ],
      "wday": "Tisdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Viltskav med smörstekta kantareller och lök, rårörda lingon samt potatismos.",
        },
        Object {
          "name": "Stekt panerad sejfilé med vitlök och saffranssmakande räkröra samt dill, citron och kokt potatis.",
        },
        Object {
          "name": "Med köttfärs, bacon, champinjoner och svartpeppar.",
          "title": "Pasta",
        },
        Object {
          "name": "Pannbiff med brynt lök, potatismos och lingonsylt samt gräddsås.",
          "title": "Husman",
        },
        Object {
          "name": "Chilibiff med rostade rotfrukter samt paprikasås.",
          "title": "Vegetariskt",
        },
        Object {
          "name": "Välj mellan kyckling eller hand skalade räkor med krutonger, bacon, parmesanost & ceasardressing.",
          "title": "Caesarsallad",
        },
      ],
      "wday": "Onsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Kokt Oxbringa med rotmos och pepparrotssås.",
        },
        Object {
          "name": "Stekt kummel med räkor, pepparrot, brynt smör och gröna ärtor serveras med kokt potatis",
        },
        Object {
          "name": "Med köttfärs, bacon, champinjoner och svartpeppar.",
          "title": "Pasta",
        },
        Object {
          "name": "Pannbiff med brynt lök, potatismos och lingonsylt samt gräddsås.",
          "title": "Husman",
        },
        Object {
          "name": "Chilibiff med rostade rotfrukter samt paprikasås.",
          "title": "Vegetariskt",
        },
        Object {
          "name": "Välj mellan kyckling eller hand skalade räkor med krutonger, bacon, parmesanost & ceasardressing.",
          "title": "Caesarsallad",
        },
      ],
      "wday": "Torsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Schnitzel med svampsås samt stekt potatis.",
        },
        Object {
          "name": "Gravad lax med dillstuvad potatis samt hovmästarsås.",
        },
        Object {
          "name": "Efterrätt ingår i alla dagens rätter.",
        },
        Object {
          "name": "Med köttfärs, bacon, champinjoner och svartpeppar.",
          "title": "Pasta",
        },
        Object {
          "name": "Pannbiff med brynt lök, potatismos och lingonsylt samt gräddsås.",
          "title": "Husman",
        },
        Object {
          "name": "Chilibiff med rostade rotfrukter samt paprikasås.",
          "title": "Vegetariskt",
        },
        Object {
          "name": "Välj mellan kyckling eller hand skalade räkor med krutonger, bacon, parmesanost & ceasardressing.",
          "title": "Caesarsallad",
        },
      ],
      "wday": "Fredag",
    },
  ],
  "name": "Hops",
  "url": "http://www.hopsbar.se/",
}
`;
